### PR TITLE
feat/interop: add conditional deployment of `SystemConfigInterop`

### PR DESCRIPTION
**Pull request by 0xfuturistic**
enables the deployment of `SystemConfigInterop` if the interop flag is set to true

### CODE REVIEW

File: packages/contracts-bedrock/scripts/ChainAssertions.sol [hunk 1]
```diff
@@ -121,6 +122,75 @@ library ChainAssertions {
...
+            require(config.scalar() == _cfg.gasPriceOracleScalar());
+            require(config.batcherHash() == bytes32(uint256(uint160(_cfg.batchSenderAddress()))));
+            require(config.gasLimit() == uint64(_cfg.l2GenesisBlockGasLimit()));
+            require(config.unsafeBlockSigner() == _cfg.p2pSequencerAddress());
+            // Check _config
+            ResourceMetering.ResourceConfig memory rconfig = Constants.DEFAULT_RESOURCE_CONFIG();
+            require(resourceConfig.maxResourceLimit == rconfig.maxResourceLimit);
+            require(resourceConfig.elasticityMultiplier == rconfig.elasticityMultiplier);
+            require(resourceConfig.baseFeeMaxChangeDenominator == rconfig.baseFeeMaxChangeDenominator);
+            require(resourceConfig.systemTxMaxGas == rconfig.systemTxMaxGas);
```
Comment (tynes) [inline packages/contracts-bedrock/scripts/ChainAssertions.sol]: Same feedback as portal PR, we don't need to introduce specific stuff, this is a copy/paste, this adds too much code. Its like a 5 loc change inside of `deploySystemConfig` that optionally deploys the interop version if its enabled

File: packages/contracts-bedrock/scripts/Deploy.s.sol [hunk 2]
```diff
@@ -1012,6 +1044,53 @@ contract Deploy is Deployer {
         ChainAssertions.checkSystemConfig({ _contracts: _proxies(), _cfg: cfg, _isProxy: true });
     }
 
+    /// @notice Initialize the SystemConfigInterop
+    function initializeSystemConfigInterop() public broadcast {
+        console.log("Upgrading and initializing SystemConfigInterop proxy");
```
Comment (tynes) [inline packages/contracts-bedrock/scripts/Deploy.s.sol]: If you every catch yourself copy/pasting huge blocks of code like this it should be a red flag to think about the problem from a diff pov

File: packages/contracts-bedrock/scripts/Deploy.s.sol [hunk 3]
```diff
@@ -784,20 +785,32 @@ contract Deploy is Deployer {
...
+        if (cfg.useInterop()) {
+            console.log("Deploying SystemConfigInterop implementation");
 
-        save("SystemConfig", address(config));
-        console.log("SystemConfig deployed at %s", address(config));
+            SystemConfigInterop config = new SystemConfigInterop{ salt: _implSalt() }();
+
+            addr_ = address(config);
+
+            save("SystemConfigInterop", addr_);
```
Comment (tynes) [inline packages/contracts-bedrock/scripts/Deploy.s.sol]: You should keep the name as if its system config for the smallest diff

```
save("SystemConfig", addr_);
```

File: packages/contracts-bedrock/scripts/Deploy.s.sol [hunk 4]
```diff
@@ -967,7 +980,14 @@ contract Deploy is Deployer {
 
     /// @notice Initialize the SystemConfig
     function initializeSystemConfig() public broadcast {
-        console.log("Upgrading and initializing SystemConfig proxy");
+        address systemConfig;
+        if (cfg.useInterop()) {
+            console.log("Upgrading and initializing SystemConfigInterop proxy");
+            systemConfig = mustGetAddress("SystemConfigInterop");
```
Comment (tynes) [inline packages/contracts-bedrock/scripts/Deploy.s.sol]: This diff is not necessary with the above

File: packages/contracts-bedrock/scripts/Deploy.s.sol [hunk 5]
```diff
@@ -1005,9 +1025,15 @@ contract Deploy is Deployer {
             )
         });
 
-        SystemConfig config = SystemConfig(systemConfigProxy);
-        string memory version = config.version();
-        console.log("SystemConfig version: %s", version);
+        if (cfg.useInterop()) {
+            SystemConfigInterop config = SystemConfigInterop(systemConfigProxy);
+            string memory version = config.version();
```
Comment (tynes) [inline packages/contracts-bedrock/scripts/Deploy.s.sol]: This diff is also not necessary there is a logline elsewhere that says its interop enabled

File: packages/contracts-bedrock/scripts/Deploy.s.sol [hunk 6]
```diff
@@ -784,20 +785,32 @@ contract Deploy is Deployer {
...
+            save("SystemConfig", addr_);
+            console.log("SystemConfigInterop deployed at %s", addr_);
+        } else {
+            console.log("Deploying SystemConfig implementation");
+
+            SystemConfig config = new SystemConfig{ salt: _implSalt() }();
+
+            addr_ = address(config);
+
+            save("SystemConfig", addr_);
```
Comment (tynes) [inline packages/contracts-bedrock/scripts/Deploy.s.sol]: You can dedupe `save` and `console.log` as they can be the same in both branches, we have a log that lets us know interop is happening already so no need to change `console.log`

Comment (0xfuturistic) [reply packages/contracts-bedrock/scripts/Deploy.s.sol:804]: for sure! forgot to push a change here removing the branch

**Review by tynes — APPROVED** [review]

**Pull request merged**